### PR TITLE
[SPARK-31319][SQL][DOCS] Document UDFs/UDAFs in SQL Reference

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Document UDF in SQL Reference

### Why are the changes needed?
To make SQL Reference complete.

### Does this PR introduce any user-facing change?
Yes. Here are the new pages:
Screen Shot 2020-04-09 at 5 06 42 PM

Screen Shot 2020-04-09 at 5 07 06 PM

Screen Shot 2020-04-09 at 5 07 26 PM

Screen Shot 2020-04-09 at 5 07 54 PM

Screen Shot 2020-04-09 at 5 08 17 PM

Screen Shot 2020-04-09 at 5 09 27 PM

Screen Shot 2020-04-09 at 5 10 00 PM

Screen Shot 2020-04-09 at 5 11 10 PM

Screen Shot 2020-04-09 at 5 11 41 PM

Screen Shot 2020-04-09 at 5 16 22 PM

Screen Shot 2020-04-09 at 5 22 15 PM

Screen Shot 2020-04-09 at 5 22 46 PM

Screen Shot 2020-04-09 at 5 23 08 PM

Screen Shot 2020-04-09 at 5 23 34 PM

Screen Shot 2020-04-09 at 5 23 57 PM

Screen Shot 2020-04-09 at 5 24 15 PM

### How was this patch tested?
Manually build and check

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @gatorsmile @maropu

File: docs/sql-ref-functions-udf-aggregate.md [hunk 1]
```diff
@@ -1,22 +1,65 @@
...
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 ---
 
-**This page is under construction**
+### Description
+
+User-Defined Aggregate Functions (UDAFs) are user-defined functions that act on multiple rows at once and return a single value as a result. This documentation contains examples that demonstrate how to define and register UDAFs in Scala and invoke them in SPARK SQL.
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:24]: nit:
 - How about `User-Defined Aggregate Functions (UDAFs) are user-defined functions` -> `... are user-programmable routines`?
 - `return a single value` -> `return a single aggregated value`?
 - `SPARK SQL` -> `Spark SQL`?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 2]
```diff
@@ -1,22 +1,65 @@
 ---
 layout: global
-title: User defined Aggregate Functions (UDAF)
-displayTitle: User defined Aggregate Functions (UDAF)
+title: User Defined Aggregate Functions (UDAF)
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:3]: nit: `UDAF` -> `UDAFs`?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 3]
```diff
@@ -1,22 +1,65 @@
...
-**This page is under construction**
+### Description
+
+User-Defined Aggregate Functions (UDAFs) are user-defined functions that act on multiple rows at once and return a single value as a result. This documentation contains examples that demonstrate how to define and register UDAFs in Scala and invoke them in SPARK SQL.
+
+### Examples
+
+{% highlight sql %}
+
+// Define, register a UDAF to calculate the sum of product of two columns
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:30]: `Define, register a UDAF to` -> `Define and register an UDAF to`>

File: docs/sql-ref-functions-udf-aggregate.md [hunk 4]
```diff
@@ -1,22 +1,65 @@
...
+// Scala
+import functions.udf
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:32]: `import functions.udf` -> `import org.apache.spark.sql.functions.udf`?

Comment (maropu) [reply docs/sql-ref-functions-udf-aggregate.md:32]: Also, we need to import `import org.apache.spark.sql.expressions.Aggregator` and `import org.apache.spark.sql.Encoder`?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 5]
```diff
@@ -1,22 +1,65 @@
...
+  (2, 4))
+  .toDF("a", "b")
+
+df.createOrReplaceTempView("testUDAF")
+
+-- SQL
+SELECT agg(a, b) FROM testUDAF;
+
++---------------------------------------------+
+|$anon$2(CAST(a AS BIGINT), CAST(b AS BIGINT))|
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:60]: btw, (this is not related to this pr though), the `"agg"` name defined in `spark.udf.register("agg", agg)` is not used in the name of the plan...

File: docs/sql-ref-functions-udf-scalar.md [hunk 6]
```diff
@@ -1,22 +1,125 @@
 ---
 layout: global
-title: User defined Scalar Functions (UDF)
-displayTitle: User defined Scalar Functions (UDF)
+title: Scalar User Defined Functions (UDF)
```
Comment (maropu) [inline docs/sql-ref-functions-udf-scalar.md:3]: nit: `UDF` -> `UDFs`?

File: docs/sql-ref-functions-udf-scalar.md [hunk 7]
```diff
@@ -1,22 +1,125 @@
...
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 ---
 
-**This page is under construction**
+### Description
+
+This documentation contains examples that demonstrate how to define and register UDFs that act on a single row and invoke them in SPARK SQL.
```
Comment (maropu) [inline docs/sql-ref-functions-udf-scalar.md:24]: How about following the same description with the UDAF one? e.g.,
```
User-Defined Functions (UDFs) are ...
```

File: docs/sql-ref-functions-udf-scalar.md [hunk 8]
```diff
@@ -1,22 +1,125 @@
...
+
+### Examples
+
+{% highlight sql %}
+
+// Define, register a zero argument non-deterministic UDF
```
Comment (maropu) [inline docs/sql-ref-functions-udf-scalar.md:30]: `Define, register` -> `Define and register`

File: docs/sql-ref-functions-udf-scalar.md [hunk 9]
```diff
@@ -1,22 +1,125 @@
...
+// Scala
+import functions.udf
+val foo = udf(() => Math.random())
+spark.udf.register("random", foo.asNondeterministic())
```
Comment (maropu) [inline docs/sql-ref-functions-udf-scalar.md:34]: How about documenting all the `asXXX` methods somewhere?

File: docs/sql-ref-functions-udf-scalar.md [hunk 10]
```diff
@@ -1,22 +1,125 @@
...
+
+This documentation contains examples that demonstrate how to define and register UDFs that act on a single row and invoke them in SPARK SQL.
+
+### Examples
+
+{% highlight sql %}
+
+// Define, register a zero argument non-deterministic UDF
+// Scala
+import functions.udf
```
Comment (maropu) [inline docs/sql-ref-functions-udf-scalar.md:32]: `import functions.udf` -> `import org.apache.spark.sql.functions.udf`?

File: docs/sql-ref-functions-udf-scalar.md [hunk 11]
```diff
@@ -1,22 +1,125 @@
...
+|9  |
++---+
+
+// UDF in a GROUP BY
+// Scala
+import functions.udf
+spark.udf.register("groupFunction", (n: Int) => { n > 10 })
+
+val df = Seq(("red", 1), ("red", 2), ("blue", 10),
+    ("green", 100), ("green", 200)).toDF("color", "value")
```
Comment (maropu) [inline docs/sql-ref-functions-udf-scalar.md:98]: It is better to match either format: a single value per line or multiple values per line https://github.com/apache/spark/pull/28087/files#diff-97b11639e031df253c371da5298980d9R44

File: docs/sql-ref-functions-udf.md [hunk 12]
```diff
@@ -1,25 +1,25 @@
...
   See the License for the specific language governing permissions and
   limitations under the License.
 ---
 
-Spark SQL is Apache Spark's module for working with structured data.
-This guide is a reference for Structured Query Language (SQL) for Apache 
-Spark. This document describes the SQL constructs supported by Spark in detail
-along with usage examples when applicable.
+User-Defined Functions (UDFs) are a feature of Spark SQL that allows users to define their own functions when the system's built-in functions are not enough to perform the desired task. To use UDFs in SPARK SQL, users must first define the function, then register the function with SPARK, and finally call the registered function. The User-Defined Functions can act on a single row or act on multiple rows at once.
+
```
Comment (maropu) [inline docs/sql-ref-functions-udf.md:23]: We don't need to document Hive UDFs here?

**Comment by gatorsmile** [comment]
Do we have a page for introducing the APIs for Scala, Java, Python UDFs/UDAFs/UDTF? If not, could we introduce it?

File: docs/sql-ref-functions-udf-aggregate.md (hunk 3 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-aggregate.md:30]: I changed the first part, but  I guess it's a before UDAF?

File: docs/sql-ref-functions-udf-scalar.md (hunk 9 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-scalar.md:34]: Do you mean add a comment  to explain asNondeterministic?

File: docs/sql-ref-functions-udf.md (hunk 12 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf.md:23]: I will think about this.

**Comment by huaxingao** [comment]
> Do we have a page for introducing the APIs for Scala, Java, Python UDFs/UDAFs/UDTF? If not, could we introduce it?

I will think about where and how to document these.

Thanks for your review! @maropu @gatorsmile

File: docs/sql-ref-functions-udf-scalar.md (hunk 9 above)
Comment (maropu) [reply docs/sql-ref-functions-udf-scalar.md:36]: I didn't mean so, but the comment looks nice.
My original suggestion wasI that, how about adding a subsection (just like `[Parameter](https://github.com/apache/spark/blob/master/docs/sql-ref-syntax-qry-select.md#parameters)`) about the methods `asXXXX`?

Comment (huaxingao) [reply docs/sql-ref-functions-udf-scalar.md:36]: I see. Since we need a page to document the UDF APIs, I will have a subsection for asXXX.

**Comment by gatorsmile** [comment]
cc @Ngone51 Please help review this?

**Comment by huaxingao** [comment]
I added Scala APIs. Java APIs and Python APIs are similar so I prefer not to add them here. I think the Scala APIs can already give the users a very good idea about what these UDF APIs are. If they need to check the Java or Python APIs, they can refer to the APIs doc. I don't want the majority of the page become API docs.

File: docs/sql-ref-functions-udf-aggregate.md [hunk 13]
```diff
@@ -21,17 +21,79 @@ license: |
...
+  </dd>
+</dl>
 
 ### Examples
 
 {% highlight sql %}
 
 // Define and register a UDAF to calculate the sum of product of two columns
 // Scala
-import org.apache.spark.sql.Encoder
```
Comment (huaxingao) [inline docs/sql-ref-functions-udf-aggregate.md]: I am confused with this import. My test works OK without it.

File: docs/sql-ref-functions-udf-aggregate.md [hunk 14]
```diff
@@ -1,22 +1,130 @@
...
+  (2, -1),
+  (4, 7),
+  (3, 8),
+  (2, 4))
+  .toDF("a", "b")
+
+df.createOrReplaceTempView("testUDAF")
+
+-- SQL
+SELECT agg(a, b) FROM testUDAF;
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:122]: If this is a Scala case, `sql("SELECT agg(a, b) FROM testUDAF").show()` is better?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 15]
```diff
@@ -1,22 +1,130 @@
...
+- IN The input type for the aggregation.
+
+- BUF The type of the intermediate value of the reduction.
+
+- OUT The type of the final output result.
+
+<dl>
+  <dt><code><em>bufferEncoder: Encoder[BUF]</em></code></dt>
+  <dd>
+    Register a deterministic Java UDF(0-22) instance as user-defined function (UDF).
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:38]: `as user-defined function (UDF).` -> `as an user-defined function (UDF).`?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 16]
```diff
@@ -1,22 +1,130 @@
...
+  <dt><code><em>outputEncoder: Encoder[OUT]</em></code></dt>
+  <dd>
+    Specifies the Encoder for the final output value type.
+  </dd>
+</dl>
+
+<dl>
+  <dt><code><em>reduce(b: BUF, a: IN): BUF</em></code></dt>
+  <dd>
+    Combine two values to produce a new value. For performance, the function may modify b and return it instead of constructing new object for b.
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:67]: `a new value` -> `a new aggregated value`?

Comment (maropu) [reply docs/sql-ref-functions-udf-aggregate.md:67]: better to wrap b with \`, e.g., \`b\`?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 17]
```diff
@@ -1,22 +1,130 @@
...
   See the License for the specific language governing permissions and
   limitations under the License.
 ---
 
-**This page is under construction**
+### Description
+
+User-Defined Aggregate Functions (UDAFs) are user-programmable routines that act on multiple rows at once and return a single aggregated value as a result. This documentation lists the classes that are required for creating and registering UDAFs. It also contains examples that demonstrate how to define and register UDAFs in Scala and invoke them in Spark SQL.
+
+### org.apache.spark.sql.expressions.Aggregator[-IN, BUF, OUT]
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:26]: How about "Interface of `Aggregator[-IN, BUF, OUT]`"? We need the full-qualifier here?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 18]
```diff
@@ -1,22 +1,130 @@
...
+</dl>
+
+<dl>
+  <dt><code><em>zero: BUF</em></code></dt>
+  <dd>
+    A zero value for this aggregation.
+  </dd>
+</dl>
+
+### org.apache.spark.sql.UDFRegistration
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:78]: ditto: how about "Inerface of `UDFRegistration`"?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 19]
```diff
@@ -1,22 +1,130 @@
...
+
+Functions for registering user-defined functions. Use `SparkSession.udf` to access this: `spark.udf`
+
+<dl>
+  <dt><code><em>register(name: String, udf: UserDefinedFunction): UserDefinedFunction</em></code></dt>
+  <dd>
+    Registers a user-defined function (UDF).
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:85]: nit: `a` -> `an`

File: docs/sql-ref-functions-udf-scalar.md [hunk 20]
```diff
@@ -1,22 +1,187 @@
...
+-- SQL
+SELECT plusOne(5);
+
++------+
+|UDF(5)|
++------+
+|     6|
++------+
+
+// Define a two arguments UDF and register it with Spark in one step
```
Comment (maropu) [inline docs/sql-ref-functions-udf-scalar.md:114]: `Define a two arguments` -> `Define two arguments`?

File: docs/sql-ref-functions-udf-aggregate.md (hunk 14 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-aggregate.md:122]: I will change all examples to this format

File: docs/sql-ref-functions-udf-aggregate.md (hunk 16 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-aggregate.md:67]: changed to <code>b</code> because ` doesn't work inside html.

File: docs/sql-ref-functions-udf-aggregate.md (hunk 17 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-aggregate.md:26]: I guess I will just use ```Aggregator[-IN, BUF, OUT]```. Because ```UserDefinedFunction```is an abstract class , and ```UDFRegistration``` is a regular class

File: docs/sql-ref-functions-udf-aggregate.md (hunk 18 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-aggregate.md:78]: changed to ```UDFRegistration```

File: docs/sql-ref-functions-udf-scalar.md (hunk 20 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-scalar.md:114]: I guess I will change to a two-argument UDF?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 21]
```diff
@@ -1,22 +1,130 @@
...
+- BUF The type of the intermediate value of the reduction.
+
+- OUT The type of the final output result.
+
+<dl>
+  <dt><code><em>bufferEncoder: Encoder[BUF]</em></code></dt>
+  <dd>
+    Register a deterministic Java UDF(0-22) instance as user-defined function (UDF).
+  </dd>
+</dl>
```
Comment (viirya) [inline docs/sql-ref-functions-udf-aggregate.md:40]: Looks not correct. Isn't `bufferEncoder` the Encoder for the intermediate value?

File: docs/sql-ref-functions-udf-aggregate.md (hunk 16 above)
Comment (viirya) [inline docs/sql-ref-functions-udf-aggregate.md]: Combine two values to produce a new value -> Aggregate input value `a` into current intermediate value?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 22]
```diff
@@ -1,22 +1,129 @@
...
+  <dt><code><em>reduce(b: BUF, a: IN): BUF</em></code></dt>
+  <dd>
+    Combine two values to produce a new aggregated value. For performance, the function may modify <code>b</code> and return it instead of constructing new object for <code>b</code>.
+  </dd>
+</dl>
+
+<dl>
+  <dt><code><em>zero: BUF</em></code></dt>
+  <dd>
+    A zero value for this aggregation.
```
Comment (viirya) [inline docs/sql-ref-functions-udf-aggregate.md:74]: The initial value of the intermediate result for this aggregation.

File: docs/sql-ref-functions-udf-aggregate.md [hunk 23]
```diff
@@ -1,22 +1,129 @@
...
+  </dd>
+</dl>
+
+### UDFRegistration
+
+Functions for registering user-defined functions. Use `SparkSession.udf` to access this: `spark.udf`
```
Comment (viirya) [inline docs/sql-ref-functions-udf-aggregate.md:80]: Use `SparkSession.udf` to access this: `spark.udf`? Sounds a bit strange.

File: docs/sql-ref-functions-udf-aggregate.md (hunk 21 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-aggregate.md:40]: Will change

File: docs/sql-ref-functions-udf-aggregate.md (hunk 16 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-aggregate.md]: I actually copied the exact API doc for this one and the next. But your text is better. I will change.

File: docs/sql-ref-functions-udf-scalar.md [hunk 24]
```diff
@@ -1,22 +1,181 @@
...
+sql("SELECT SUM(value) FROM groupData GROUP BY groupFunction(value)").show()
+
++----------+
+|sum(value)|
++----------+
+|13        |
+|300       |
++----------+
+
+# Define and register a UDF using Python
```
Comment (viirya) [inline docs/sql-ref-functions-udf-scalar.md:167]: Do we need this section? I think there is doc for Python UDFs. No?

File: docs/sql-ref-functions-udf-aggregate.md (hunk 23 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-aggregate.md:80]: `SparkSession.udf` to access this: `spark.udf` -> `SparkSession.udf` to access this.
Is this better?

Comment (viirya) [reply docs/sql-ref-functions-udf-aggregate.md:80]: Yea, I just wonder why `spark.udf` is there.

File: docs/sql-ref-functions-udf-scalar.md (hunk 24 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-scalar.md:167]: Just want to give an example using Python. There is API doc for Python UDFs, but I won't have a separate doc for Python. I am OK to keep or delete it.

**Comment by huaxingao** [comment]
@srowen Could you please check this one as well?

File: docs/sql-ref-functions-udf-scalar.md [hunk 25]
```diff
@@ -1,22 +1,181 @@
...
+
+<dl>
+  <dt><code><em>deterministic: Boolean</em></code></dt>
+  <dd>
+    Returns true iff the UDF is deterministic, i.e. the UDF produces the same output given the same input.
+  </dd>
+</dl>
+
+<dl>
+  <dt><code><em>nullable: Boolean</em></code></dt>
```
Comment (maropu) [inline docs/sql-ref-functions-udf-scalar.md:53]: we need `deterministic` and `nullable` here? I personally think `asNonNullable`, `asNondeterministic`, and `withName ` is enough for users.

File: docs/sql-ref-functions-udf-scalar.md [hunk 26]
```diff
@@ -1,22 +1,181 @@
...
 
-**This page is under construction**
+### Description
+
+User-Defined Functions (UDFs) are user-programmable routines that act on one row. This documentation lists the classes that are required for creating and registering UDFs. It also contains examples that demonstrate how to define and register UDFs and invoke them in Spark SQL.
+
+
+### UserDefinedFunction
+
+A user-defined function. To create one, use the `udf` functions in `functions`.
```
Comment (maropu) [inline docs/sql-ref-functions-udf-scalar.md:29]: How about saying here `To define the properties of a user-defined function, you can use the some methods defined in 'UserDefinedFunction';`?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 27]
```diff
@@ -1,22 +1,133 @@
...
+<dl>
+  <dt><code><em>zero: BUF</em></code></dt>
+  <dd>
+    The initial value of the intermediate result for this aggregation.
+  </dd>
+</dl>
+
+### UDFRegistration
+
+Functions for registering user-defined functions. Use `SparkSession.udf` to access this.
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:80]: Users need to know this class? I personally think not, so how about removing this section?

File: docs/sql-ref-functions-udf-aggregate.md [hunk 28]
```diff
@@ -1,22 +1,133 @@
...
 
-**This page is under construction**
+### Description
+
+User-Defined Aggregate Functions (UDAFs) are user-programmable routines that act on multiple rows at once and return a single aggregated value as a result. This documentation lists the classes that are required for creating and registering UDAFs. It also contains examples that demonstrate how to define and register UDAFs in Scala and invoke them in Spark SQL.
+
+### Aggregator[-IN, BUF, OUT]
+
+A base class for user-defined aggregations, which can be used in Dataset operations to take all of the elements of a group and reduce them to a single value.
+- IN The input type for the aggregation.
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:29]: nit: need a single blank?
```
A base class for user-defined aggregations, which can be used in Dataset operations to take all of the elements of a group and reduce them to a single value.

- IN The input type for the aggregation.
```

Comment (maropu) [reply docs/sql-ref-functions-udf-aggregate.md:29]: nit: also, how about putting a separator between `IN` and the statement? e.g., `IN - The input...`

Comment (maropu) [reply docs/sql-ref-functions-udf-aggregate.md:29]: btw, the contents are somewhat overlapped with https://people.apache.org/~pwendell/spark-nightly/spark-master-docs/latest/sql-programming-guide.html#type-safe-user-defined-aggregate-functions ?

Comment (maropu) [reply docs/sql-ref-functions-udf-aggregate.md:29]: btw, the contents are somewhat overlapped with https://spark.apache.org/docs/3.0.0-preview/sql-getting-started.html#aggregations ? Probably, we need to organize these aggregator contents.

File: docs/sql-ref-functions-udf-aggregate.md (hunk 27 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-aggregate.md:80]: user needs to call ```UDFRegistration.register``` to register the udaf 
```spark.udf.register("agg", agg)```
I guess we keep this section?

File: docs/sql-ref-functions-udf-aggregate.md (hunk 28 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-aggregate.md:29]: I will think about this.

Comment (maropu) [reply docs/sql-ref-functions-udf-aggregate.md]: I personally think its better to move all the contents of `sql-getting-started.html` into the SQL doc. Since the user-defined aggregator topic is advanced one, the SQL doc is more suitable than the `Getting Started` doc. cc: @gatorsmile

File: docs/sql-ref-functions-udf-aggregate.md [hunk 29]
```diff
@@ -1,22 +1,134 @@
...
+### Examples
+
+{% highlight sql %}
+
+// Define and register a UDAF to calculate the sum of product of two columns
+// Scala
+import org.apache.spark.sql.expressions.Aggregator
+import org.apache.spark.sql.functions.udaf
+
+val agg = udaf(new Aggregator[(Long, Long), Long, Long] {
```
Comment (Ngone51) [inline docs/sql-ref-functions-udf-aggregate.md:99]: Seems we already have an example at https://spark.apache.org/docs/latest/sql-getting-started.html#aggregations.

And do we still need to keep that `Aggregations` section or not? If not, I think we could reuse the previous examples here.

File: docs/sql-ref-functions-udf-aggregate.md [hunk 30]
```diff
@@ -1,22 +1,134 @@
...
+    Registers a user-defined function (UDF).
+  </dd>
+</dl>
+
+### Examples
+
+{% highlight sql %}
+
+// Define and register a UDAF to calculate the sum of product of two columns
+// Scala
```
Comment (Ngone51) [inline docs/sql-ref-functions-udf-aggregate.md:95]: Can we highlight Scala codes as other doc does? e.g,
```
<div data-lang="scala"  markdown="1">
{% include_example generic_load_save_functions scala/org/apache/spark/examples/sql/SQLDataSourceExample.scala %}
</div>
```

File: docs/sql-ref-functions-udf-aggregate.md (hunk 28 above)
Comment (gatorsmile) [reply docs/sql-ref-functions-udf-aggregate.md]: @maropu sounds good to me.

File: docs/sql-ref-functions-udf-aggregate.md (hunk 27 above)
Comment (maropu) [reply docs/sql-ref-functions-udf-aggregate.md:81]: I still think its ok for users to just know how to register udf/udaf in the document, so user don't need to know what `UDFRegistration` is.

File: examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedScalar.scala [hunk 31]
```diff
@@ -0,0 +1,81 @@
...
+    // |  7|
+    // |  8|
+    // |  9|
+    // +---+
+
+    // $example off:udf_scalar$
+
+    spark.stop()
+  }
+
```
Comment (maropu) [inline examples/src/main/scala/org/apache/spark/examples/sql/UserDefinedScalar.scala:80]: nit: remove the  blank line.

File: docs/sql-ref-functions-udf-scalar.md [hunk 32]
```diff
@@ -1,22 +1,77 @@
...
+<div class="codetabs">
+<div data-lang="scala"  markdown="1">
+{% include_example udf_scalar scala/org/apache/spark/examples/sql/UserDefinedScalar.scala%}
+</div>
+<div data-lang="java"  markdown="1">
+{% include_example udf_scalar java/org/apache/spark/examples/sql/JavaUserDefinedScalar.java%}
+</div>
+</div>
+
+
```
Comment (maropu) [inline docs/sql-ref-functions-udf-scalar.md:74]: nit: remove the single blank line.

File: docs/sql-ref-functions-udf-scalar.md [hunk 33]
```diff
@@ -1,22 +1,77 @@
...
   See the License for the specific language governing permissions and
   limitations under the License.
 ---
 
-**This page is under construction**
+### Description
+
+User-Defined Functions (UDFs) are user-programmable routines that act on one row. This documentation lists the classes that are required for creating and registering UDFs. It also contains examples that demonstrate how to define and register UDFs and invoke them in Spark SQL.
+
+
```
Comment (maropu) [inline docs/sql-ref-functions-udf-scalar.md:26]: nit: remove the blank line.

File: docs/sql-ref-functions-udf-aggregate.md [hunk 34]
```diff
@@ -1,22 +1,121 @@
...
+<div data-lang="scala"  markdown="1">
+{% include_example untyped_custom_aggregation scala/org/apache/spark/examples/sql/UserDefinedUntypedAggregation.scala%}
+</div>
+<div data-lang="java"  markdown="1">
+{% include_example untyped_custom_aggregation java/org/apache/spark/examples/sql/JavaUserDefinedUntypedAggregation.java%}
+</div>
+</div>
+
+### Related Statements
+- [Scalar User Defined Functions (UDFs)](sql-ref-functions-udf-scalar.html)
```
Comment (maropu) [inline docs/sql-ref-functions-udf-aggregate.md:120]: nit: `-` -> `*`

**Review by maropu — APPROVED** [review]
LGTM except for the existing comments.

File: docs/sql-ref-functions-udf-aggregate.md (hunk 27 above)
Comment (huaxingao) [reply docs/sql-ref-functions-udf-aggregate.md:73]: I will remove

**Comment by srowen** [comment]
Merged to master/3.0

**Comment by huaxingao** [comment]
Thank you everyone!

File: examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedScalar.java [hunk 35]
```diff
@@ -0,0 +1,98 @@
...
+    // +----------+
+    // |         6|
+    // +----------+
+
+    // Define and register a two-argument UDF
+    UserDefinedFunction strLen = udf(
+      (String s, Integer x) -> s.length() + x, DataTypes.IntegerType
+    );
+    spark.udf().register("strLen", strLen);
+    spark.sql("SELECT strLen('test', 1)").show();
```
Comment (HyukjinKwon) [inline examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedScalar.java:70]: It should have been better probably if we added SQL example with UDAF too, e.g.)

```sql
val agg = Aggregator(...)
spark.udf.register("udaf_func", agg)
```

which was added in SPARK-27296.

The current example is only Scala/Java APIs there in `UserDefinedTypedAggregation.scala` and `JavaUserDefinedTypedAggregation.java`.

Comment (maropu) [reply examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedScalar.java:70]: +1; @huaxingao Could you add it in follow-up? The initali commit in this PR include the example (https://github.com/apache/spark/pull/28087/commits/b276e16df4c0543b30583cb2007b31cdc78a0848#diff-97b11639e031df253c371da5298980d9R32-R54), but it seems to disappear during the reviews....

Comment (huaxingao) [reply examples/src/main/java/org/apache/spark/examples/sql/JavaUserDefinedScalar.java:70]: I just submitted a follow-up to add the SQL example.
